include/cufinufft/impl.h

```diff
@@ -0,0 +1,503 @@
+#ifndef CUFINUFFT_IMPL_H
+#define CUFINUFFT_IMPL_H
+
+#include <iostream>
+
+#include <helper_cuda.h>
+
+#include <cufinufft/common.h>
+#include <cufinufft/cudeconvolve.h>
+#include <cufinufft/defs.h>
+#include <cufinufft/memtransfer.h>
+#include <cufinufft/spreadinterp.h>
+#include <cufinufft/types.h>
+#include <cufinufft/utils.h>
+
+// 1d
+template <typename T>
+int cufinufft1d1_exec(cuda_complex<T> *d_c, cuda_complex<T> *d_fk, cufinufft_plan_t<T> *d_plan);
+template <typename T>
+int cufinufft1d2_exec(cuda_complex<T> *d_c, cuda_complex<T> *d_fk, cufinufft_plan_t<T> *d_plan);
+
+// 2d
+template <typename T>
+int cufinufft2d1_exec(cuda_complex<T> *d_c, cuda_complex<T> *d_fk, cufinufft_plan_t<T> *d_plan);
+template <typename T>
+int cufinufft2d2_exec(cuda_complex<T> *d_c, cuda_complex<T> *d_fk, cufinufft_plan_t<T> *d_plan);
+
+// 3d
+template <typename T>
+int cufinufft3d1_exec(cuda_complex<T> *d_c, cuda_complex<T> *d_fk, cufinufft_plan_t<T> *d_plan);
+template <typename T>
+int cufinufft3d2_exec(cuda_complex<T> *d_c, cuda_complex<T> *d_fk, cufinufft_plan_t<T> *d_plan);
+
+static void cufinufft_setup_binsize(int type, int dim, cufinufft_opts *opts) {
+    switch (dim) {
+    case 1: {
+        opts->gpu_binsizex = (opts->gpu_binsizex < 0) ? 1024 : opts->gpu_binsizex;
+        opts->gpu_binsizey = 1;
+        opts->gpu_binsizez = 1;
+    } break;
+    case 2: {
+        opts->gpu_binsizex = (opts->gpu_binsizex < 0) ? 32 : opts->gpu_binsizex;
+        opts->gpu_binsizey = (opts->gpu_binsizey < 0) ? 32 : opts->gpu_binsizey;
+        opts->gpu_binsizez = 1;
+    } break;
+    case 3: {
+        switch (opts->gpu_method) {
+        case 1:
+        case 2: {
+            opts->gpu_binsizex = (opts->gpu_binsizex < 0) ? 16 : opts->gpu_binsizex;
+            opts->gpu_binsizey = (opts->gpu_binsizey < 0) ? 16 : opts->gpu_binsizey;
+            opts->gpu_binsizez = (opts->gpu_binsizez < 0) ? 2 : opts->gpu_binsizez;
+        } break;
+        case 4: {
+            opts->gpu_obinsizex = (opts->gpu_obinsizex < 0) ? 8 : opts->gpu_obinsizex;
+            opts->gpu_obinsizey = (opts->gpu_obinsizey < 0) ? 8 : opts->gpu_obinsizey;
+            opts->gpu_obinsizez = (opts->gpu_obinsizez < 0) ? 8 : opts->gpu_obinsizez;
+            opts->gpu_binsizex = (opts->gpu_binsizex < 0) ? 4 : opts->gpu_binsizex;
+            opts->gpu_binsizey = (opts->gpu_binsizey < 0) ? 4 : opts->gpu_binsizey;
+            opts->gpu_binsizez = (opts->gpu_binsizez < 0) ? 4 : opts->gpu_binsizez;
+        } break;
+        }
+    } break;
+    }
+}
+
+template <typename T>
```

Comment:
I didn't originally have it implemented like this, but it was difficult to prevent the compiler from optimizing the internal calls away, leaving the instantiated templates to not be exported to the shared object. This allows me to call the implementations in tests without worrying about symbol exports.

Reply:
Ah ok. So it's there to simplify testing, that's all?

Reply:
Pretty much, though it's not an uncommon way to implement lots of header-only type code.

Reply:
Ok.